src/config.c

```diff
@@ -1678,14 +1678,21 @@ int rewriteConfigOverwriteFile(char *configfile, sds content) {
         serverLog(LL_WARNING, "Could not chmod config file (%s)", strerror(errno));
     else if (rename(tmp_conffile, configfile) == -1)
         serverLog(LL_WARNING, "Could not rename tmp config file (%s)", strerror(errno));
+    else if (fsyncFileDir(configfile) == -1)
+        serverLog(LL_WARNING, "Could not sync config file dir (%s)", strerror(errno));
     else {
         retval = 0;
         serverLog(LL_DEBUG, "Rewritten config file (%s) successfully", configfile);
     }
 
 cleanup:
     close(fd);
-    if (retval) unlink(tmp_conffile);
+    if (retval) {
+        /* try to unlink the temp, ignoring possible errors */
+        int old_errno = errno;
```

Comment:
why do we backup `errno` here? we already printed the error and errno, and the caller doesn't need it.
if ti did, then we should back it up before `close`

Reply:
I thought a bad directory `fsync` would make the state of temp file uncertain, so the `unlink` may fail and corrupt the errno. Your advise wrapping before `close` is more universal, so I refactor into this way.